# An analysis of CAO point from 2019, 2020 and 2021

---

In [1]:
# Regular expressions
import re

# For HTTP requests
import requests as rq

# Dates and Times
import datetime as dt

# Pandas
import pandas as pd

# For saving urls
import urllib.request as urlrq

# FOr only printing top few lines of large datasets
from itertools import islice

In [2]:
# Get current date and time
now = dt.datetime.now()

# Format as string
nowstr = now.strftime('%Y%m%d_%H%M%S')

***

# 2021 CAO Points

https://www.cao.ie/index.php?page=points&p=2021

***

## Note on HTML format for 2021 CAO Points

At the time of starting this project, the CAO points for 2021 were only available in on a HTML webpage. The link above would have another link taking you to this webpage. Presently, that link now links to a .xlsx file. Although the web address for the old HTML webpage of 2021 points still works, it cannot be accessed from the main 2021 page linked to above. While the .xlsx will presumably be updated with any new points, the same cannot be said of the HTML page. I have left the code to access the HTML webpage in but for the analysis of the data, I will be using the data from the .xlsx file.

### Level 8 Courses (HTML)

In [3]:
resp = rq.get('http://www2.cao.ie/points/l8.php')

# Check if working
# Response [200] if working
resp

<Response [200]>

In [4]:
# Compile the regular expression for matching lines

# This compiler has some issues and does not get all lines needed
# A simpler compiler might be possible
# re_course = re.compile('([A-Z]{2}[0-9]{3})  (.*)([0-9]{3})(\*?) *')

re_course = re.compile('([A-Z]{2}[0-9]{3}.*)')



In [5]:
# Loop through line of the response content
no_lines = 0
for line in resp.iter_lines():
    dline = line.decode('cp1252')
    # Only match lines with courses / points
    if re_course.fullmatch(dline):
        no_lines +=1
        # print(line)
        csv_version = re_course.sub(r'\1', dline)
        # print(csv_version)
print(f"----------------------------------------------------------------------------- \n Total number of lines is {no_lines}.")

----------------------------------------------------------------------------- 
 Total number of lines is 949.


In [6]:
# Create file path for saving csv file
path = 'data/cao2021_l8' + nowstr + '.csv'

# Keep track of number of lines
no_lines = 0

with open(path, 'w') as f:
    
    for line in resp.iter_lines():
        dline = line.decode('cp1252')
        if re_course.fullmatch(dline):
            no_lines = no_lines + 1
            linesplit = re.split('  +', dline)
            f.write(','.join(linesplit) + '\n')
        
print(f"Total number of lines is {no_lines}.")

Total number of lines is 949.


### Level 6/7 Courses (HTML)

In [7]:
resp = rq.get('http://www2.cao.ie/points/l76.php')

# Check if working
# Response [200] if working
resp

<Response [200]>

In [8]:
# Compile the regular expression for matching lines

# This compiler has some issues and does not get all lines needed
# A simpler compiler might be possible
# re_course = re.compile('([A-Z]{2}[0-9]{3})  (.*)([0-9]{3})(\*?) *')

re_course = re.compile('([A-Z]{2}[0-9]{3}.*)')



In [9]:
# Loop through line of the response content
no_lines = 0
for line in resp.iter_lines():
    dline = line.decode('cp1252')
    # Only match lines with courses / points
    if re_course.fullmatch(dline):
        no_lines +=1
        # print(line)
        csv_version = re_course.sub(r'\1', dline)
        # print(csv_version)
print(f"----------------------------------------------------------------------------- \n Total number of lines is {no_lines}.")

----------------------------------------------------------------------------- 
 Total number of lines is 416.


In [10]:
# Create file path for saving csv file
path = 'data/cao2021_l76' + nowstr + '.csv'

# Keep track of number of lines
no_lines = 0

with open(path, 'w') as f:
    
    for line in resp.iter_lines():
        dline = line.decode('cp1252')
        if re_course.fullmatch(dline):
            no_lines = no_lines + 1
            linesplit = re.split('  +', dline)
            f.write(','.join(linesplit) + '\n')
        
print(f"Total number of lines is {no_lines}.")

Total number of lines is 416.


### Level 6/7/8 Courses (xlsx)

In [11]:
# Name new excel save files
path = 'data/cao2021_' + nowstr + '.xlsx'

In [12]:
# Save Excel File
urlrq.urlretrieve('http://www2.cao.ie/points/CAOPointsCharts2021.xlsx', path)

('data/cao2021_20220101_170726.xlsx',
 <http.client.HTTPMessage at 0x25d0fb99f40>)

In [13]:
# Download and parse the excel spreadsheet
df2021 = pd.read_excel('http://www2.cao.ie/points/CAOPointsCharts2021.xlsx', skiprows=11)

In [14]:
# Save Pandas dataframe to file
path = 'data/cao2021_' + nowstr + '.csv'

df2021.to_csv(path)

In [15]:
# Filter columns that we want to see
df2021 = df2021[["Course Code", "Course Title", "EOS Points", "EOS Midpoints", "Course Level"]]

In [16]:
 columns = ["Code", "Title", "Points", "MidPoints", "Level"]

In [17]:
df2021.columns = columns

In [18]:
df2021.head()

Code                            Title Points MidPoints  Level
0  AL605  Music and Instrument Technology    211       319      6
1  AL630              Pharmacy Technician    308       409      6
2  AL631                   Dental Nursing    311       400      6
3  AL632                  Applied Science    297       454      6
4  AL650                         Business    AQA       351      6

***

# 2020 CAO Points

https://www.cao.ie/index.php?page=points&p=2020

***

In [19]:
# Name new excel save files
path = 'data/cao2020_' + nowstr + '.xlsx'

In [20]:
# Save Excel File
urlrq.urlretrieve('http://www2.cao.ie/points/CAOPointsCharts2020.xlsx', path)

('data/cao2020_20220101_170726.xlsx',
 <http.client.HTTPMessage at 0x25d21333550>)

In [21]:
# Download and parse the excel spreadsheet
df2020 = pd.read_excel('http://www2.cao.ie/points/CAOPointsCharts2020.xlsx', skiprows=10)

In [22]:
# Save Pandas dataframe to file
path = 'data/cao2020_' + nowstr + '.csv'

df2020.to_csv(path)

In [23]:
# Filter columns that we want to see
df2020 = df2020[["COURSE CODE2", "COURSE TITLE", "EOS", "EOS Mid-point", "LEVEL"]]

In [24]:
df2020.columns = columns

In [25]:
df2020.head()

Code                                              Title    Points  \
0  AC120                             International Business       209   
1  AC137                                       Liberal Arts       252   
2  AD101   First Year Art & Design (Common Entry,portfolio)  #+matric   
3  AD102  Graphic Design and Moving Image Design (portfo...  #+matric   
4  AD103  Textile & Surface Design and Jewellery & Objec...  #+matric   

  MidPoints  Level  
0       280      8  
1       270      8  
2  #+matric      8  
3  #+matric      8  
4  #+matric      8

***

# 2019 CAO Points

https://www.cao.ie/index.php?page=points&p=2019

***

### Steps to reproduce

1. Download original pdf file
2. Open original pdf file in Microsoft Word
3. Save Microsoft Word's converted pdf in docx format
4. Re-save Word document for editing
5. Delete headers and footers
6. Delete preamble on page 1
7. Select all and copy
8. Paste into Excel
9. Manually delete blank lines
10. Save as xlsx document

### Level 8 Courses

In [26]:
# assign name for saving data
path = 'data/cao2019_l8' + nowstr + '.tsv'

In [27]:
# Convert downlaoded excel to dataframe
df2019L8 = pd.read_excel('data/cao2019_20211102_200630.xlsx')

In [28]:
# Save as csv
df2019L8.to_csv(path, sep= '\t')

In [29]:
# Rename columns to match other dfs
df2019L8.columns = columns[:-1]

In [30]:
df2019L8["Level"] = "8"

In [31]:
df2019L8.head()

Code                                              Title Points  MidPoints  \
0  AL801    Software Design with Virtual Reality and Gaming    304      328.0   
1  AL802               Software Design with Cloud Computing    301      306.0   
2  AL803  Software Design with Mobile Apps and Connected...    309      337.0   
3  AL805        Network Management and Cloud Infrastructure    329      442.0   
4  AL810                                 Quantity Surveying    307      349.0   

  Level  
0     8  
1     8  
2     8  
3     8  
4     8

### Level 6/7 Courses

In [32]:
# assign name for saving data
path = 'data/cao2019_l67' + nowstr + '.tsv'

In [33]:
# Convert downlaoded excel to dataframe
df2019L67 = pd.read_excel('data/cao2019_20211202_193530_l67.xlsx')

In [34]:
# Save as csv
df2019L67.to_csv(path, sep= '\t')

In [35]:
# Rename columns to match other dfs
df2019L67.columns = columns[:-1]

In [36]:
df2019L67["Level"] = "6/7"

In [37]:
df2019L67.head()

Code                   Title Points  MidPoints Level
0  AL600         Software Design    205      306.0   6/7
1  AL601    Computer Engineering    196      272.0   6/7
2  AL602  Mechanical Engineering    258      424.0   6/7
3  AL604       Civil Engineering    252      360.0   6/7
4  AL630     Pharmacy Technician    306      366.0   6/7

## Joining of Years

In [38]:
# Append level 6/7s to level 8s
df2019 = df2019L8.append(df2019L67)

In [39]:
df2019L8.count()

Code         930
Title        930
Points       926
MidPoints    915
Level        930
dtype: int64

In [40]:
df2019L67.count()

Code         461
Title        461
Points       454
MidPoints    453
Level        461
dtype: int64

In [41]:
df2019.count()

Code         1391
Title        1391
Points       1380
MidPoints    1368
Level        1391
dtype: int64

## Joining of Dataframes

In [42]:
# Assign different column names in preparation for merging
columns2021 = ["Code", "Title", "2021Points", "2021MidPoints", "2021Level"]
columns2020 = ["Code", "Title", "2020Points", "2020MidPoints", "2020Level"]
columns2019 = ["Code", "Title", "2019Points", "2019MidPoints", "2019Level"]

In [43]:
df2021.columns = columns2021
df2020.columns = columns2020
df2019.columns = columns2019

In [44]:
df2021.head()

Code                            Title 2021Points 2021MidPoints  2021Level
0  AL605  Music and Instrument Technology        211           319          6
1  AL630              Pharmacy Technician        308           409          6
2  AL631                   Dental Nursing        311           400          6
3  AL632                  Applied Science        297           454          6
4  AL650                         Business        AQA           351          6

In [45]:
df2020.head()

Code                                              Title 2020Points  \
0  AC120                             International Business        209   
1  AC137                                       Liberal Arts        252   
2  AD101   First Year Art & Design (Common Entry,portfolio)   #+matric   
3  AD102  Graphic Design and Moving Image Design (portfo...   #+matric   
4  AD103  Textile & Surface Design and Jewellery & Objec...   #+matric   

  2020MidPoints  2020Level  
0           280          8  
1           270          8  
2      #+matric          8  
3      #+matric          8  
4      #+matric          8

In [46]:
df2019.head()

Code                                              Title 2019Points  \
0  AL801    Software Design with Virtual Reality and Gaming        304   
1  AL802               Software Design with Cloud Computing        301   
2  AL803  Software Design with Mobile Apps and Connected...        309   
3  AL805        Network Management and Cloud Infrastructure        329   
4  AL810                                 Quantity Surveying        307   

   2019MidPoints 2019Level  
0          328.0         8  
1          306.0         8  
2          337.0         8  
3          442.0         8  
4          349.0         8

In [47]:
dfcao = pd.merge(df2021, df2020, on="Code", how="outer")

In [48]:
dfcao.count()

Code             1558
Title_x          1451
2021Points       1422
2021MidPoints    1433
2021Level        1451
Title_y          1464
2020Points       1448
2020MidPoints    1437
2020Level        1464
dtype: int64

In [49]:
dfcao = pd.merge(dfcao, df2019, on = "Code", how = "outer")

In [50]:
dfcao.count()

Code             1812
Title_x          1451
2021Points       1422
2021MidPoints    1433
2021Level        1451
Title_y          1464
2020Points       1448
2020MidPoints    1437
2020Level        1464
Title            1391
2019Points       1380
2019MidPoints    1368
2019Level        1391
dtype: int64

### Removing NaNs
For the analysis of the points across the years, means etc will be thrown off if we include some courses that are present in some years and not in others. We will create a dataframe that only has courses that were present across all 3 years. We will also have to drop any courses that have non-number points as these cannot be analysed. Likewise for courses requiring portfolios (denoted with a # in the dataset) as we do not know the graded of these portfolios.

In [51]:
dfcao = dfcao.dropna()

In [52]:
# search for rows not containing AQA and # in selected columns
# ~ returns the opposite of the search
dfcao = dfcao[~dfcao['2021Points'].str.contains("AQA", na=False)]
dfcao = dfcao[~dfcao['2020Points'].str.contains("AQA", na=False)]
dfcao = dfcao[~dfcao['2019Points'].str.contains("AQA", na=False)]

dfcao = dfcao[~dfcao['2021Points'].str.contains("#", na=False)]
dfcao = dfcao[~dfcao['2020Points'].str.contains("#", na=False)]
dfcao = dfcao[~dfcao['2019Points'].str.contains("#", na=False)]

In [53]:
dfcao.count()

Code             936
Title_x          936
2021Points       936
2021MidPoints    936
2021Level        936
Title_y          936
2020Points       936
2020MidPoints    936
2020Level        936
Title            936
2019Points       936
2019MidPoints    936
2019Level        936
dtype: int64

In [54]:
dfcao = dfcao.drop(["Title_y", "Title"], axis = 1)

In [55]:
dfcao

Code                                          Title_x 2021Points  \
1     AL630                              Pharmacy Technician        308   
2     AL631                                   Dental Nursing        311   
3     AL632                                  Applied Science        297   
8     AL701  Computer Engineering for Network Infrastructure        207   
11    AL704                             Computer Engineering        210   
...     ...                                              ...        ...   
1444  WD208                        Manufacturing Engineering        317   
1445  WD210                     Software Systems Development        260   
1446  WD211                               Creative Computing        270   
1447  WD212                  Recreation and Sport Management        262   
1448  WD230         Mechanical and Manufacturing Engineering        230   

     2021MidPoints  2021Level 2020Points 2020MidPoints  2020Level 2019Points  \
1              409        6.0        295           379        6.0        306   
2              400        6.0        298           388        6.0        326   
3              454        6.0        234           434        6.0        243   
8              382        7.0        216           337        7.0        235   
11             368        7.0        204           403        7.0        205   
...            ...        ...        ...           ...        ...        ...   
1444           368        7.0        188           339        7.0        195   
1445           373        8.0        279           337        8.0        271   
1446           392        8.0        271           318        8.0        275   
1447           304        8.0        270           349        8.0        274   
1448           361        8.0        253           369        8.0        273   

      2019MidPoints 2019Level  
1             366.0       6/7  
2             379.0       6/7  
3             372.0       6/7  
8             310.0       6/7  
11            307.0       6/7  
...             ...       ...  
1444          330.0       6/7  
1445          329.0         8  
1446          322.0         8  
1447          311.0         8  
1448          348.0         8  

[936 rows x 11 columns]

In [56]:
 #Enabling this option allows for a scan of all rows in dataframe
 pd.options.display.max_rows = None

In [57]:
# Check for different levels in courses across years
dfcao.loc[(dfcao["2021Level"] == 8.0) & (dfcao["2020Level"] != 8.0)]

Empty DataFrame
Columns: [Code, Title_x, 2021Points, 2021MidPoints, 2021Level, 2020Points, 2020MidPoints, 2020Level, 2019Points, 2019MidPoints, 2019Level]
Index: []

In [58]:
dfcao.loc[(dfcao["2021Level"] == 8.0) & (dfcao["2019Level"] != "8")]

Empty DataFrame
Columns: [Code, Title_x, 2021Points, 2021MidPoints, 2021Level, 2020Points, 2020MidPoints, 2020Level, 2019Points, 2019MidPoints, 2019Level]
Index: []

In [59]:
dfcao.loc[(dfcao["2021Level"] == 7.0) & (dfcao["2020Level"] != 7.0)]

Empty DataFrame
Columns: [Code, Title_x, 2021Points, 2021MidPoints, 2021Level, 2020Points, 2020MidPoints, 2020Level, 2019Points, 2019MidPoints, 2019Level]
Index: []

In [60]:
dfcao.loc[(dfcao["2021Level"] == 6.0) & (dfcao["2020Level"] != 6.0)]

Empty DataFrame
Columns: [Code, Title_x, 2021Points, 2021MidPoints, 2021Level, 2020Points, 2020MidPoints, 2020Level, 2019Points, 2019MidPoints, 2019Level]
Index: []

In [61]:
dfcao.loc[(dfcao["2021Level"] == 7.0) & (dfcao["2019Level"] != "6/7")]

Empty DataFrame
Columns: [Code, Title_x, 2021Points, 2021MidPoints, 2021Level, 2020Points, 2020MidPoints, 2020Level, 2019Points, 2019MidPoints, 2019Level]
Index: []

In [62]:
dfcao.loc[(dfcao["2021Level"] == 6.0) & (dfcao["2019Level"] != "6/7")]

Empty DataFrame
Columns: [Code, Title_x, 2021Points, 2021MidPoints, 2021Level, 2020Points, 2020MidPoints, 2020Level, 2019Points, 2019MidPoints, 2019Level]
Index: []

Because all course levels are the same across the years, we can remove the 2020 and 2019 levels.

In [63]:
#Remove 2020 and 2019 levels
dfcao = dfcao.drop(["2020Level", "2019Level"], axis = 1)

In [64]:
dfcao.head()

Code                                          Title_x 2021Points  \
1   AL630                              Pharmacy Technician        308   
2   AL631                                   Dental Nursing        311   
3   AL632                                  Applied Science        297   
8   AL701  Computer Engineering for Network Infrastructure        207   
11  AL704                             Computer Engineering        210   

   2021MidPoints  2021Level 2020Points 2020MidPoints 2019Points  2019MidPoints  
1            409        6.0        295           379        306          366.0  
2            400        6.0        298           388        326          379.0  
3            454        6.0        234           434        243          372.0  
8            382        7.0        216           337        235          310.0  
11           368        7.0        204           403        205          307.0

In [65]:
# Rename title and level columns
dfcao = dfcao.rename(columns = {"Title_x": "Title", "2021Level": "Level"})

In [66]:
dfcao.head()

Code                                            Title 2021Points  \
1   AL630                              Pharmacy Technician        308   
2   AL631                                   Dental Nursing        311   
3   AL632                                  Applied Science        297   
8   AL701  Computer Engineering for Network Infrastructure        207   
11  AL704                             Computer Engineering        210   

   2021MidPoints  Level 2020Points 2020MidPoints 2019Points  2019MidPoints  
1            409    6.0        295           379        306          366.0  
2            400    6.0        298           388        326          379.0  
3            454    6.0        234           434        243          372.0  
8            382    7.0        216           337        235          310.0  
11           368    7.0        204           403        205          307.0

In [67]:
# Rearrange columns
dfcao = dfcao[["Code", "Title","Level","2021Points","2021MidPoints","2020Points","2020MidPoints","2019Points","2019MidPoints"]]

In [68]:
dfcao.head()

Code                                            Title  Level 2021Points  \
1   AL630                              Pharmacy Technician    6.0        308   
2   AL631                                   Dental Nursing    6.0        311   
3   AL632                                  Applied Science    6.0        297   
8   AL701  Computer Engineering for Network Infrastructure    7.0        207   
11  AL704                             Computer Engineering    7.0        210   

   2021MidPoints 2020Points 2020MidPoints 2019Points  2019MidPoints  
1            409        295           379        306          366.0  
2            400        298           388        326          379.0  
3            454        234           434        243          372.0  
8            382        216           337        235          310.0  
11           368        204           403        205          307.0

In [69]:
# Convert 2019 Points to string before removing * (* means not all points got on course)
dfcao["2019Points"] = dfcao["2019Points"].astype(str)
dfcao['2019Points'] = dfcao['2019Points'].str.replace(r"\*",'')

In [70]:
dfcao.head()

Code                                            Title  Level 2021Points  \
1   AL630                              Pharmacy Technician    6.0        308   
2   AL631                                   Dental Nursing    6.0        311   
3   AL632                                  Applied Science    6.0        297   
8   AL701  Computer Engineering for Network Infrastructure    7.0        207   
11  AL704                             Computer Engineering    7.0        210   

   2021MidPoints 2020Points 2020MidPoints 2019Points  2019MidPoints  
1            409        295           379        306          366.0  
2            400        298           388        326          379.0  
3            454        234           434        243          372.0  
8            382        216           337        235          310.0  
11           368        204           403        205          307.0

In [71]:
# Convert Points columns to ints
dfcao["2021Points"] = pd.to_numeric(dfcao["2021Points"])
dfcao["2020Points"] = pd.to_numeric(dfcao["2020Points"])
dfcao["2019Points"] = pd.to_numeric(dfcao["2019Points"])

In [72]:
print(dfcao[["Title", "2021Points"]].loc[dfcao["2021Points"] > 625], "\n\n")
print(dfcao[["Title", "2020Points"]].loc[dfcao["2020Points"] > 625], "\n\n")
print(dfcao[["Title", "2019Points"]].loc[dfcao["2019Points"] > 625], "\n\n")

                                                  Title  2021Points
875                        Contemporary Art (portfolio)         775
879                        Contemporary Art (portfolio)         790
1362  Animation, Visual Effects and Motion Design (L...         718 


                            Title  2020Points
875  Contemporary Art (portfolio)         702
879  Contemporary Art (portfolio)         679 


Empty DataFrame
Columns: [Title, 2019Points]
Index: [] 




From the above results of points over 625, it appears that 3 courses have portfolios in their points grading but were not marked as such. These will be removed.

In [73]:
dfcao = dfcao.drop([875,879,1362])

In [74]:
print(dfcao[["Title", "2021Points"]].loc[dfcao["2021Points"] > 625], "\n\n")
print(dfcao[["Title", "2020Points"]].loc[dfcao["2020Points"] > 625], "\n\n")
print(dfcao[["Title", "2019Points"]].loc[dfcao["2019Points"] > 625], "\n\n")

Empty DataFrame
Columns: [Title, 2021Points]
Index: [] 


Empty DataFrame
Columns: [Title, 2020Points]
Index: [] 


Empty DataFrame
Columns: [Title, 2019Points]
Index: [] 




In [75]:
dfcao["2021Diff"] = dfcao["2021Points"] - dfcao["2020Points"]
dfcao["2020Diff"] = dfcao["2020Points"] - dfcao["2019Points"]

In [76]:
dfcao.head()

Code                                            Title  Level  2021Points  \
1   AL630                              Pharmacy Technician    6.0         308   
2   AL631                                   Dental Nursing    6.0         311   
3   AL632                                  Applied Science    6.0         297   
8   AL701  Computer Engineering for Network Infrastructure    7.0         207   
11  AL704                             Computer Engineering    7.0         210   

   2021MidPoints  2020Points 2020MidPoints  2019Points  2019MidPoints  \
1            409         295           379         306          366.0   
2            400         298           388         326          379.0   
3            454         234           434         243          372.0   
8            382         216           337         235          310.0   
11           368         204           403         205          307.0   

    2021Diff  2020Diff  
1         13       -11  
2         13       -28  
3         63        -9  
8         -9       -19  
11         6        -1

# Analysis of CAO Points

In [82]:
dfcao.count()

Code             933
Title            933
Level            933
2021Points       933
2021MidPoints    933
2020Points       933
2020MidPoints    933
2019Points       933
2019MidPoints    933
2021Diff         933
2020Diff         933
dtype: int64

#### Total number of Courses Analysed
In total, 933 course were analysed out of a potential 1812 courses. The majority of the courses not analysed were removed because they were not present across all 3 years. Other courses were removed due to some points data not being present or they courses requiring portfolios which data was not given for.

In [96]:
dfcao.describe()

Level  2021Points  2020Points  2019Points  2019MidPoints  \
count  933.000000  933.000000  933.000000  933.000000     933.000000   
mean     7.658092  349.286174  335.355841  324.513398     392.229368   
std      0.574803  122.874533  110.829277  105.503123      79.604865   
min      6.000000  100.000000   80.000000  101.000000     190.000000   
25%      7.000000  260.000000  252.000000  249.000000     330.000000   
50%      8.000000  315.000000  309.000000  306.000000     374.000000   
75%      8.000000  444.000000  419.000000  401.000000     445.000000   
max      8.000000  625.000000  613.000000  601.000000     613.000000   

         2021Diff    2020Diff  
count  933.000000  933.000000  
mean    13.930332   10.842444  
std     34.063694   29.150156  
min   -181.000000 -157.000000  
25%     -1.000000   -2.000000  
50%     12.000000    9.000000  
75%     32.000000   25.000000  
max    238.000000  188.000000

In [88]:
# Separate courses by level
dfcaoL6 = dfcao.loc[dfcao["Level"] == 6]
dfcaoL7 = dfcao.loc[dfcao["Level"] == 7]
dfcaoL8 = dfcao.loc[dfcao["Level"] == 8]

In [97]:
dfcaoL8.describe()

Level  2021Points  2020Points  2019Points  2019MidPoints    2021Diff  \
count  663.0  663.000000  663.000000  663.000000     663.000000  663.000000   
mean     8.0  394.508296  377.693816  365.895928     410.114630   16.814480   
std      0.0  110.662150   98.780608   91.380102      81.420946   30.189387   
min      8.0  176.000000  195.000000  168.000000     262.000000 -181.000000   
25%      8.0  301.000000  300.000000  298.000000     346.000000    0.000000   
50%      8.0  384.000000  363.000000  349.000000     393.000000   18.000000   
75%      8.0  488.000000  454.500000  434.000000     476.000000   34.000000   
max      8.0  625.000000  613.000000  601.000000     613.000000  101.000000   

         2020Diff  
count  663.000000  
mean    11.797888  
std     25.839527  
min    -99.000000  
25%      0.000000  
50%     11.000000  
75%     25.000000  
max    148.000000

In [98]:
dfcaoL7.describe()

Level  2021Points  2020Points  2019Points  2019MidPoints    2021Diff  \
count  221.0  221.000000  221.000000  221.000000     221.000000  221.000000   
mean     7.0  240.126697  232.579186  223.692308     351.067873    7.547511   
std      0.0   64.403573   56.659263   55.873114      51.247793   37.770404   
min      7.0  100.000000   80.000000  101.000000     255.000000 -154.000000   
25%      7.0  200.000000  195.000000  186.000000     311.000000   -8.000000   
50%      7.0  234.000000  233.000000  222.000000     347.000000    8.000000   
75%      7.0  280.000000  269.000000  260.000000     382.000000   20.000000   
max      7.0  433.000000  400.000000  393.000000     520.000000  136.000000   

         2020Diff  
count  221.000000  
mean     8.886878  
std     35.931258  
min   -154.000000  
25%     -8.000000  
50%      3.000000  
75%     21.000000  
max    188.000000

In [99]:
dfcaoL6.describe()

Level  2021Points  2020Points  2019Points  2019MidPoints    2021Diff  \
count   49.0   49.000000   49.000000   49.000000      49.000000   49.000000   
mean     6.0  229.734694  226.040816  219.306122     335.877551    3.693878   
std      0.0   86.895957   60.152223   69.984881      64.140027   54.882528   
min      6.0  113.000000  148.000000  116.000000     190.000000  -81.000000   
25%      6.0  171.000000  181.000000  173.000000     297.000000  -24.000000   
50%      6.0  214.000000  210.000000  206.000000     329.000000   -1.000000   
75%      6.0  259.000000  251.000000  245.000000     377.000000   11.000000   
max      6.0  463.000000  451.000000  454.000000     499.000000  238.000000   

         2020Diff  
count   49.000000  
mean     6.734694  
std     36.410722  
min   -157.000000  
25%     -9.000000  
50%      9.000000  
75%     29.000000  
max     82.000000

---
# End